txscript/pkscript.go

```diff
@@ -0,0 +1,194 @@
+package txscript
+
+import (
+	"crypto/sha256"
+	"errors"
+	"fmt"
+
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+	"golang.org/x/crypto/ripemd160"
+)
+
+const (
+	// pubKeyHashSigScriptLen is the length of a signature script attempting
+	// to spend a P2PKH script. This length is determined by the following:
+	//   0x48 (72 byte data push) | <71 byte sig> | <1 byte sighash flag> |
+	//   0x21 (33 byte data push) | <33 byte compressed pubkey>
+	pubKeyHashSigScriptLen = 107
+
+	// witnessV0PubKeyHashLen is the length of a P2WPKH script.
+	witnessV0PubKeyHashLen = 22
+
+	// scriptHashLen is the length of a P2SH script.
+	scriptHashLen = 23
+
+	// witnessV0ScriptHashLen is the length of a P2WSH script.
+	witnessV0ScriptHashLen = 34
+
+	// maxLen is the maximum script length supported by ParsePkScript.
+	maxLen = witnessV0ScriptHashLen
+)
+
+var (
+	// ErrUnsupportedScriptType ...
+	ErrUnsupportedScriptType = errors.New("unsupported script type")
+)
+
+// PkScript is a wrapper struct around a byte array, allowing it to be used
+// as a map index.
+type PkScript struct {
```

Comment:
We _could_ also extend this to cover a set of arbitrary scripts, but given that for our particular use case, we only care about segwit style scripts, I think we can defer that. However, since this only covers `v0` witness programs, I think we should update the name to reflect that into `txscript.WitnessProgram`.

Reply:
Would say the P2SH stuff falls into this bucket as well, since it's really just a witness program with a layer of indirection.

Reply:
Decided to keep as is after our discussion offline. We'll now also support P2PKH scripts.